clang/lib/CodeGen/CGDebugInfo.cpp

```diff
@@ -1332,6 +1332,54 @@ llvm::DIType *CGDebugInfo::CreateType(const TemplateSpecializationType *Ty,
   auto PP = getPrintingPolicy();
   Ty->getTemplateName().print(OS, PP, TemplateName::Qualified::None);
 
+  SourceLocation Loc = AliasDecl->getLocation();
+
+  if (CGM.getCodeGenOpts().DebugTemplateAlias) {
+    // TemplateSpecializationType doesn't know if its template args are
+    // being substituted into a parameter pack. We can find out if that's
+    // the case now by inspecting the TypeAliasTemplateDecl template
+    // parameters. Insert Ty's template args into SpecArgs, bundling args
+    // passed to a parameter pack into a TemplateArgument::Pack.
+    SmallVector<TemplateArgument> SpecArgs;
+    {
+      ArrayRef SubstArgs = Ty->template_arguments();
+      for (const NamedDecl *P : TD->getTemplateParameters()->asArray()) {
+        if (P->isParameterPack()) {
+          SpecArgs.push_back(TemplateArgument(SubstArgs));
+          break;
+        }
+        // Skip defaulted args.
+        if (SubstArgs.empty()) {
+          // If SubstArgs is now empty (we're taking from it each iteration) and
+          // this template parameter isn't a pack, then that should mean we're
+          // using default values for the remaining template parameters.
+          break;
```

Comment:
Does this mean we won't be emitting the defaulted template arguments as part of the `DW_TAG_template_alias`? Would be good to add a test case for defaulted arguments, so this logic is exercised (if you haven't already)

Also minor nit, is this logic getting sufficiently long to warrant a separate helper function?

Reply:
Ah I see, thanks for clarifying! Just thought I'd point it out in case this wasn't expected. The DWARF spec has this to say about the template parameters of an alias:
```
The template alias entry has child entries describing the template actual parameters (see Section 2.23 on page 57).
```
Which to me implies that the parameters of an alias instantiation follow the same rules as those for class template instantiations.

Regarding name reconstruction, defaulted template arguments should have their `IsDefaulted` bit set, so when reconstructing the name we should be able to drop defaulted arguments? If that's too much of a hassle maybe not emitting them at all as you do now is also fine?

Reply:
I'm glad you pointed it out - it was doing what I wanted it to but I hadn't tried very hard to get it to gather default argument values, and I agree copying the class template instantiation debug info is the safest bet.

I've implemented that now. Slightly worried there are other template parameter kinds I've not thought of, currently guarded by an llvm_unrecahable. Maybe there's something better to do there - do you have any thoughts on this?

Reply:
Ah - this implementation appears to cause an assertion when one of the template parameters is a dependant type... (working on it)

Reply:
> lightly worried there are other template parameter kinds I've not thought of, currently guarded by an llvm_unrecahable. Maybe there's something better to do there - do you have any thoughts on this?

The three kinds you check for should be sufficient afaik. Places around clang consistently only deal with these three types of parameter decls.

Reply:
> The three kinds you check for should be sufficient afaik. Places around clang consistently only deal with these three types of parameter decls.

Excellent, that is reassuring, thanks.

I'm wondering if you've got any ideas about my dependent value situation:

```
template <int I1 = 5, int I2 = I1>
using B = int;
B<> b;
```

The current default-argument-gathering tactic can't cope with dependent default expressions/types. Do you know if there's any way that I can "resolve", for lack of a better word, the template parameter default expression to get an expression `= 5` rather than the dependent `= I1`? Presumably this is doable, because we've got a specialisation that defines what `I1` is (are there cases where this could never work... maybe with SFINAE? I'm not sure).

I'm continuing to look into it, I'm only asking in case you know off the top of your head, no problem if not.

Reply:
Clang doesn't resolve defaultedness for dependent arguments either. As a consequence, when a template parameter has a dependent default argument then its `DW_TAG_template_value_parameter` won't have a `DW_AT_default_value`. So I don't think we need to handle this if it's going to be hard to do. Can we just arrange for the parameter to be emitted as a non-defaulted `DW_TAG_template_value_parameter`?

Reply:
Looking at a class template instantiation with a default dependent value:
```
template <int I1 = 5, int I2 = I1>
struct B {};
B<> b;
```
We get:
```
0x00000029:   DW_TAG_structure_type
                DW_AT_calling_convention        (DW_CC_pass_by_value)
                DW_AT_name      ("B<5, 5>")
                DW_AT_byte_size (0x01)
                DW_AT_decl_file ("/home/och/scratch/test11.cpp")
                DW_AT_decl_line (2)

0x0000002f:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I1")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x00000036:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I2")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x0000003d:     NULL
 ```
 
Is your suggestion to emit something like this for the template alias case (from my previous comment)?
```
0x00000029:   DW_TAG_template_alias
                DW_AT_name      ("B<5, ?>")
                ...

0x0000002f:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I1")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x00000036:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I2")

0x0000003d:     NULL
```

I think that is doable but I'm not sure how that would interact with name reconstruction (in Clang debug info without -gsimple-template-names, and in debuggers/tools with it).

tyvm for your ongoing help with this.

Reply:
Ah apologies! I must've misremembered, I could've sworn there were cases where we don't handle the defaulted-ness.

> Is your suggestion to emit something like this for the template alias case (from my previous comment)?
Yup, would default values ever matter for name reconstruction? The only thing I know of that uses `DW_AT_default_value` in LLDB is when formatting the typename of a variable when printing it (we omit defaulted template parameters).

Reply:
I suppose you could call `EvaluateKnownConstInt` (on the TemplateArgument expression, if it's `!isValueDependent() && isIntegerConstantExpr()`, similar to what `templateArgumentExpressionsEqual` does). Maybe we can do this from within `CollectTemplateParams` where it asserts that the expression is a Constant?

I'm not actually sure how we get to the `TemplateArgument::Expression` case in `CollectTemplateParams` for constant expressions. AFAICT in the AST such expressions get evaluated into `TemplateArgument::Integral` before we ever get here. Seems like you're tripping over the assert now because you're calling `CollectTemplateParams` on the arguments to a `TemplateAliasDecl` which don't have their arguments evaluated in the front-end it looks like.

Reply:
(defaulted values... and types*)

Reply:
> I suppose you could call EvaluateKnownConstInt (on the TemplateArgument expression, if it's !isValueDependent() && isIntegerConstantExpr(), similar to what templateArgumentExpressionsEqual does). Maybe we can do this from within CollectTemplateParams where it asserts that the expression is a Constant?

The constant value default appears to be working, it's the dependent value `int I2 = I1` that is causing problems for me.

> Seems like you're tripping over the assert now because you're calling CollectTemplateParams on the arguments to a TemplateAliasDecl which don't have their arguments evaluated in the front-end it looks like.

That sounds like it might be the problem. It feels like we have all the parts needed to instantiate the dependent expression at this point... (because all the types/values it depends on exist as non-default argument values), I'm just not sure how to ask Clang to do that.


Here's an example using type parameters rather than values (you can also do `typename C = X<B>` if you want to hit yet another assertion...)
```
template <typename B = int, typename C = B>
using A = int;
A<> a;
```

```
Dependent types cannot show up in debug information
UNREACHABLE executed at clang/lib/CodeGen/CGDebugInfo.cpp:3686!
...
CGDebugInfo::CreateTypeNode(clang::QualType, llvm::DIFile*)
CGDebugInfo::getOrCreateType(clang::QualType, llvm::DIFile*) 
CollectTemplateParams # from the TemplateArgument::Type case processing the type of the argument for C
```

Reply:
>Sorry for diverting you down this path, maybe ignoring the default arguments as you did in your first attempt is sufficient if this is just not feasible.

Not at all, I think it was worth exploring and I very much appreciate the detailed review and the help that's come with it.

I agree that ignoring default arguments might be the way forward then. It isn't a regression in terms of name reconstruction, since the existing typedef approach for template aliases omits defaulted arguments from the DW_AT_name anyway:

```
template <typename B = int, typename C = B>
using A = int;
A<> a;
```
`$ clang test.cpp -g -o - | llvm-dwarfdump -o -`
```
[...]
0x0000002e:   DW_TAG_typedef
                DW_AT_type      (0x00000036 "int")
                DW_AT_name      ("A<>")  <-- no defaulted arguments for template aliases :-(
                DW_AT_decl_file ("test.cpp")
                DW_AT_decl_line (41)
```

I have pushed that change (to omit defaulted arguments). Thanks for bearing with me on this one!

I've chatted to our debugger folks and they're ok with omitting defaulted args (though ideally this would be fixed in the future).

I think moving this discussion onto discourse might be useful - I'll look at doing that later in the week or next week.

Reply:
Sounds good! FWIW, LLDB would currently omit the defaulted parameters too, by choice, since usually they're just noise (especially with many of the STL types).